neqo-transport/src/recv_stream.rs

```diff
@@ -1146,18 +1249,511 @@ mod tests {
+    }
+
+    // Test flow control in RecvStreamState::DataRecvd
+    #[test]
+    fn fc_state_data_recv() {
+        const SW: u64 = 1024;
+        const SW_US: usize = 1024;
+        let fc = Rc::new(RefCell::new(ReceiverFlowControl::new((), SW)));
+
+        let mut s = create_stream_with_fc(Rc::clone(&fc), SW);
+
+        check_fc_values(&fc, &s, 0, 0, 0, 0);
+
+        s.inbound_stream_frame(true, 0, &[0; SW_US / 2]).unwrap();
+        check_fc_values(&fc, &s, SW / 2, 0, SW / 2, 0);
+
+        // Receiving duplicate frames (already consumed data) will not cause an error or
+        // change fc.
+        s.inbound_stream_frame(true, SW / 4, &[0; SW_US / 4])
+            .unwrap();
+        check_fc_values(&fc, &s, SW / 2, 0, SW / 2, 0);
+
+        // The stream can still receive duplicate data without a fin bit.
+        s.inbound_stream_frame(false, SW / 4, &[0; SW_US / 4])
+            .unwrap();
+        check_fc_values(&fc, &s, SW / 2, 0, SW / 2, 0);
+
+        // Receiving frame pass the final size of a stream will return an error.
+        assert_eq!(
+            s.inbound_stream_frame(true, SW / 4, &[0; SW_US / 4 + 1]),
+            Err(Error::FinalSizeError)
+        );
+        check_fc_values(&fc, &s, SW / 2, 0, SW / 2, 0);
+
+        // Read all data
+        let mut buf = [1; SW_US];
+        assert_eq!(s.read(&mut buf).unwrap(), (SW_US / 2, true));
+        // the stream does not have fc any more. We can only check the session fc.
+        assert_eq!(fc.borrow().consumed(), SW / 2);
+        assert_eq!(fc.borrow().retired(), SW / 2);
+    }
+
+    // Test flow control in RecvStreamState::DataRead
+    #[test]
+    fn fc_state_data_read() {
+        const SW: u64 = 1024;
+        const SW_US: usize = 1024;
+        let fc = Rc::new(RefCell::new(ReceiverFlowControl::new((), SW)));
+
+        let mut s = create_stream_with_fc(Rc::clone(&fc), SW * 3 / 4);
+        check_fc_values(&fc, &s, 0, 0, 0, 0);
+
+        s.inbound_stream_frame(true, 0, &[0; SW_US / 2]).unwrap();
+        check_fc_values(&fc, &s, SW / 2, 0, SW / 2, 0);
+
+        let mut buf = [1; SW_US];
+        assert_eq!(s.read(&mut buf).unwrap(), (SW_US / 2, true));
+        // the stream does not have fc any more. We can only check the session fc.
+        assert_eq!(fc.borrow().consumed(), SW / 2);
+        assert_eq!(fc.borrow().retired(), SW / 2);
+
+        // Receiving duplicate frames (already consumed data) will not cause an error or
+        // change fc.
+        s.inbound_stream_frame(true, 0, &[0; SW_US / 2]).unwrap();
+        // the stream does not have fc any more. We can only check the session fc.
+        assert_eq!(fc.borrow().consumed(), SW / 2);
+        assert_eq!(fc.borrow().retired(), SW / 2);
+
+        // Receiving frame pass the final size of a stream will NOT return an error.
+        s.inbound_stream_frame(true, 0, &[0; SW_US / 2 + 1])
+            .unwrap();
+
+        // Receiving frame pass the stream's fc limit will not return an error.
```

Comment:
```suggestion
        // Receiving frame past the stream's fc limit will not return an error.
```
you could just combine the comments though.